### ISSUE TITLE
heap-buffer-overflow

### ISSUE THREAD

**Issue by charlesxsh** [facebookexperimental/starlark-rust#24]
For given input file in the zip file, the code compiled with AddressSanitizer reported heap-buffer-overflow error during runtime.
[in.zip](https://github.com/facebookexperimental/starlark-rust/files/6818393/in.zip)

```rust
use starlark::eval::Evaluator;
use starlark::environment::{Module, Globals};
use starlark::values::Value;
use starlark::syntax::{AstModule, Dialect};


fn main() {
    // change filepath to the input file
    let data = std::fs::read(filepath).unwrap();
        let _  = match std::str::from_utf8(&data) {
            Ok(d) => {
                if let Ok(ast) = AstModule::parse("hello_world.star", d.to_owned(), &Dialect::Standard) {
                   
                    let globals: Globals = Globals::standard();

                    let module: Module = Module::new();

                    let mut eval: Evaluator = Evaluator::new(&module, &globals);

                    eval.eval_module(ast);  
                }

                
            },
            Err(..) => return,
        };
       
}
```

Output

```
=================================================================
==23694==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x6020000011b0 at pc 0x557edcf14bde bp 0x7ffccc795d70 sp 0x7ffccc795d68
READ of size 8 at 0x6020000011b0 thread T0
    #0 0x557edcf14bdd in starlark::values::fast_string::skip_at_most_1byte::f::h0e1da77a9859a9a1 /home/xsh/.cargo/registry/src/github.com-1ecc6299db9ec823/starlark-0.4.0/src/values/fast_string.rs:65:35
    #1 0x557edcf14255 in starlark::values::fast_string::skip_at_most_1byte::h40f70214e41aab2e /home/xsh/.cargo/registry/src/github.com-1ecc6299db9ec823/starlark-0.4.0/src/values/fast_string.rs:84:14
    #2 0x557edcf14f94 in starlark::values::fast_string::at::h317771503d053ab2 /home/xsh/.cargo/registry/src/github.com-1ecc6299db9ec823/starlark-0.4.0/src/values/fast_string.rs:89:13
    #3 0x557edd106f9e in starlark::values::types::string::_$LT$impl$u20$starlark..values..traits..StarlarkValue$u20$for$u20$alloc..boxed..Box$LT$str$GT$$GT$::at::hd2466a6c38857b75 /home/xsh/.cargo/registry/src/github.com-1ecc6299db9ec823/starlark-0.4.0/src/values/types/string.rs:180:27
    #4 0x557edd1498b2 in starlark::values::_$LT$impl$u20$starlark..values..layout..value..Value$GT$::at::h86d1c57ef797f103 /home/xsh/.cargo/registry/src/github.com-1ecc6299db9ec823/starlark-0.4.0/src/values/mod.rs:454:9
    #5 0x557edcdc9d9b in starlark::eval::expr::_$LT$impl$u20$starlark..eval..Compiler$GT$::expr::_$u7b$$u7b$closure$u7d$$u7d$::h97225333fb9eecfe /home/xsh/.cargo/registry/src/github.com-1ecc6299db9ec823/starlark-0.4.0/src/eval/expr.rs:434:25
    #6 0x557edcb3bb3c in _$LT$alloc..boxed..Box$LT$F$GT$$u20$as$u20$core..ops..function..Fn$LT$A$GT$$GT$::call::hfb7414494fc44b22 /home/xsh/code/rust-compiler/src/liballoc/boxed.rs:1095:9
    #7 0x557edce16f27 in starlark::eval::stmt::_$LT$impl$u20$starlark..eval..Compiler$GT$::stmt::_$u7b$$u7b$closure$u7d$$u7d$::h823a805583856ead /home/xsh/.cargo/registry/src/github.com-1ecc6299db9ec823/starlark-0.4.0/src/eval/stmt.rs:423:21
    #8 0x557edcb3bb3c in _$LT$alloc..boxed..Box$LT$F$GT$$u20$as$u20$core..ops..function..Fn$LT$A$GT$$GT$::call::hfb7414494fc44b22 /home/xsh/code/rust-compiler/src/liballoc/boxed.rs:1095:9
    #9 0x557edce437e1 in starlark::eval::_$LT$impl$u20$starlark..eval..context..Evaluator$GT$::eval_module::h1885c10571a8c91a /home/xsh/.cargo/registry/src/github.com-1ecc6299db9ec823/starlark-0.4.0/src/eval/mod.rs:168:19
    #10 0x557edd409ddb in starlarkfuzzvrf::main::hfa9f9430d0214a86 /home/xsh/code/rust-asan/fuzz/starlark-vrf.rs:30:21
    #11 0x557edd409986 in std::rt::lang_start::_$u7b$$u7b$closure$u7d$$u7d$::hf7e2a6fbfb396fba /home/xsh/code/rust-compiler/src/libstd/rt.rs:67:34
    #12 0x557edd966612 in std::rt::lang_start_internal::h32d9533fcd93e198 (/home/xsh/code/rust-asan/fuzz/target/shihao/x86_64-unknown-linux-gnu/debug/starlarkfuzzvrf+0x1690612)
    #13 0x557edd409ac5 in std::rt::lang_start::hddcbf92bd8879287 /home/xsh/code/rust-compiler/src/libstd/rt.rs:67:5



0x6020000011b7 is located 0 bytes to the right of 7-byte region [0x6020000011b0,0x6020000011b7)
allocated by thread T0 here:
    #0 0x557edc521cf0 in malloc /home/xsh/code/rust-compiler/src/llvm-project/compiler-rt/lib/asan/asan_malloc_linux.cpp:145
    #1 0x557edd97d721 in __rdl_alloc (/home/xsh/code/rust-asan/fuzz/target/shihao/x86_64-unknown-linux-gnu/debug/starlarkfuzzvrf+0x16a7721)
    #2 0x557edd65a220 in _$LT$alloc..alloc..Global$u20$as$u20$core..alloc..AllocRef$GT$::alloc::hcad9d1152b08d2f0 /home/xsh/code/rust-compiler/src/liballoc/alloc.rs:174:49
    #3 0x557edd69ad99 in alloc::raw_vec::RawVec$LT$T$C$A$GT$::allocate_in::hf0e3c52b7f9b2ee2 /home/xsh/code/rust-compiler/src/liballoc/raw_vec.rs:183:32
    #4 0x557edd69a81f in alloc::raw_vec::RawVec$LT$T$C$A$GT$::with_capacity_in::h90673edd5154220e /home/xsh/code/rust-compiler/src/liballoc/raw_vec.rs:159:9
    #5 0x557edcb7fd46 in _$LT$alloc..boxed..Box$LT$str$GT$$u20$as$u20$core..convert..From$LT$$RF$str$GT$$GT$::from::hab6c369a76dfc01e /home/xsh/code/rust-compiler/src/liballoc/boxed.rs:834:44
    #6 0x557edd10b562 in starlark::values::types::string::_$LT$impl$u20$starlark..values..AllocFrozenValue$u20$for$u20$$RF$str$GT$::alloc_frozen_value::h0320a8d115d99a71 /home/xsh/.cargo/registry/src/github.com-1ecc6299db9ec823/starlark-0.4.0/src/values/types/string.rs:298:24
    #7 0x557edd12ab76 in starlark::values::_$LT$impl$u20$starlark..values..layout..heap..FrozenHeap$GT$::alloc::h0f81ce735feb37a2 /home/xsh/.cargo/registry/src/github.com-1ecc6299db9ec823/starlark-0.4.0/src/values/mod.rs:143:9
    #8 0x557edcdaf5ae in starlark::eval::expr::_$LT$impl$u20$starlark..syntax..ast..AstLiteral$GT$::compile::hc9b70e02f9b69980 /home/xsh/.cargo/registry/src/github.com-1ecc6299db9ec823/starlark-0.4.0/src/eval/expr.rs:181:45
    #9 0x557edcdb700b in starlark::eval::expr::_$LT$impl$u20$starlark..eval..Compiler$GT$::expr::h85a66507d2df6b39 /home/xsh/.cargo/registry/src/github.com-1ecc6299db9ec823/starlark-0.4.0/src/eval/expr.rs:430:29
    #10 0x557edce0ec2c in starlark::eval::stmt::_$LT$impl$u20$starlark..eval..Compiler$GT$::stmt::hf456df1940516028 /home/xsh/.cargo/registry/src/github.com-1ecc6299db9ec823/starlark-0.4.0/src/eval/stmt.rs:420:25
    #11 0x557edce16317 in starlark::eval::stmt::_$LT$impl$u20$starlark..eval..Compiler$GT$::stmt::_$u7b$$u7b$closure$u7d$$u7d$::hd630830d91d07722 /home/xsh/.cargo/registry/src/github.com-1ecc6299db9ec823/starlark-0.4.0/src/eval/stmt.rs:406:52
    #12 0x557edc82cf32 in core::iter::adapters::map_fold::_$u7b$$u7b$closure$u7d$$u7d$::h2dcb1f051bc7decd /home/xsh/code/rust-compiler/src/libcore/iter/adapters/mod.rs:833:28
    #13 0x557edc7e15fa in core::iter::traits::iterator::Iterator::fold::hf811279acb3524f2 /home/xsh/code/rust-compiler/src/libcore/iter/traits/iterator.rs:2022:21
    #14 0x557edcc08dcf in _$LT$core..iter..adapters..Map$LT$I$C$F$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$::fold::h6574cb8d59d2ee82 /home/xsh/code/rust-compiler/src/libcore/iter/adapters/mod.rs:873:9
    #15 0x557edc7fecf9 in core::iter::traits::iterator::Iterator::for_each::he221ccad8ce7ce7d /home/xsh/code/rust-compiler/src/libcore/iter/traits/iterator.rs:658:9
    #16 0x557edcb47a1f in _$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$alloc..vec..SpecExtend$LT$T$C$I$GT$$GT$::spec_extend::h6d9bbc281d6dd098 /home/xsh/code/rust-compiler/src/liballoc/vec.rs:2140:17
    #17 0x557edcb5bbc7 in _$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$alloc..vec..SpecExtend$LT$T$C$I$GT$$GT$::from_iter::h2aa07f3389d4a90c /home/xsh/code/rust-compiler/src/liballoc/vec.rs:2120:9
    #18 0x557edcbf2db1 in _$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$core..iter..traits..collect..FromIterator$LT$T$GT$$GT$::from_iter::h17f17a4a8c594510 /home/xsh/code/rust-compiler/src/liballoc/vec.rs:1995:9
    #19 0x557edc7eaff7 in core::iter::traits::iterator::Iterator::collect::hc9377029bb0e0da2 /home/xsh/code/rust-compiler/src/libcore/iter/traits/iterator.rs:1671:9
    #20 0x557edca4bcf4 in _$LT$alloc..vec..Vec$LT$T$GT$$u20$as$u20$gazebo..ext..vec..VecExt$GT$::into_map::h241f89a4238fcadd /home/xsh/.cargo/registry/src/github.com-1ecc6299db9ec823/gazebo-0.2.2/src/ext/vec.rs:170:9
    #21 0x557edce0e7b7 in starlark::eval::stmt::_$LT$impl$u20$starlark..eval..Compiler$GT$::stmt::hf456df1940516028 /home/xsh/.cargo/registry/src/github.com-1ecc6299db9ec823/starlark-0.4.0/src/eval/stmt.rs:406:33
    #22 0x557edce42f67 in starlark::eval::_$LT$impl$u20$starlark..eval..context..Evaluator$GT$::eval_module::h1885c10571a8c91a 

SUMMARY: AddressSanitizer: heap-buffer-overflow /home/xsh/.cargo/registry/src/github.com-1ecc6299db9ec823/starlark-0.4.0/src/values/fast_string.rs:65:35 in starlark::values::fast_string::skip_at_most_1byte::f::h0e1da77a9859a9a1
Shadow bytes around the buggy address:
  0x0c047fff81e0: fa fa 07 fa fa fa 07 fa fa fa 06 fa fa fa 03 fa
  0x0c047fff81f0: fa fa 02 fa fa fa 02 fa fa fa 06 fa fa fa 03 fa
  0x0c047fff8200: fa fa 05 fa fa fa 02 fa fa fa 02 fa fa fa 05 fa
  0x0c047fff8210: fa fa 05 fa fa fa 04 fa fa fa 02 fa fa fa 02 fa
  0x0c047fff8220: fa fa 04 fa fa fa 03 fa fa fa 04 fa fa fa 03 fa
=>0x0c047fff8230: fa fa 04 fa fa fa[07]fa fa fa 00 fa fa fa 00 fa
  0x0c047fff8240: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c047fff8250: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c047fff8260: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c047fff8270: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c047fff8280: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==23694==ABORTING
```

**Comment by ndmitchell** [comment]
Thanks for the report - investigating now! Any hints on how you found this issue?

**Comment by charlesxsh** [comment]
No problem at all. I am an PhD student from PSU and currently we are working on a fuzzing project aiming to find rust bugs effectively.

**Comment by ndmitchell** [comment]
Thanks very much for taking the time to fuzz Starlark - we much appreciate. I confirm you found a genuine bug. The fix is in c3328fe83b70d416bb307c9dbbb298155ac621e4, and then I followed up with a better fix and debug assertion for the specific issue at 320d54c3ca14aadd709b5841318ced7062f17bd5.

If you're planning to do any future fuzz testing work on Starlark, we'd love to know the results, and be happy to provide any details that would be useful for your academic work/papers.
